eval/eval.py

```diff
@@ -608,9 +1201,85 @@ def output_token_scores() -> None:
             write_basic_value(wrong_span, "wrong_span", num_tokens, tp)
             calc_recall(right_span, wrong_span, "right_span", "wrong_span", "spanrecall")
 
-        output_duration()
-        output_sentence_scores()
-        output_token_scores()
+        def output_error_cat_scores() -> None:
+            """ Calculate and write scores for each error category to stdout"""
+            print("\n\nResults for each error category in order by frequency")
+            freqdict = defaultdict(float)
+            micro : float = 0.0
+            nfreqs : int  = 0
+            microall : float = 0.0
+            nfreqsall : int = 0
+
+            # Iterate over category counts
+            for cat in self._errtypes.keys():
+                # Get recall, precision and F1; recall for correction and span
+                calc_error_category_metrics(cat)
+
+                # Collect  micro scores, both overall and for in-scope categories
+                if cat not in OUT_OF_SCOPE:
+                    micro += self._errtypes[cat]["fscore"]*self._errtypes[cat]["freq"]
+                    nfreqs += self._errtypes[cat]["freq"]
+                microall += self._errtypes[cat]["fscore"]*self._errtypes[cat]["freq"]
+                nfreqsall += self._errtypes[cat]["freq"]
+
+                # Create freqdict for sorting error categories by frequency
+                freqdict[cat] = self._errtypes[cat]["freq"]
+
+            # print results for each category by frequency
+            for k in sorted(freqdict, key=freqdict.get, reverse=True):
+                print("{} (in_scope={})".format(k, k not in OUT_OF_SCOPE))
+                print("\tTP, FP, FN: {}, {}, {}".format(self._errtypes[k]["tp"], self._errtypes[k]["fp"], self._errtypes[k]["fn"]))
+                print("\tRe, Pr, F1: {:3.2f}, {:3.2f}, {:3.2f}".format(self._errtypes[k]["recall"]*100.0, self._errtypes[k]["precision"]*100.0, self._errtypes[k]["fscore"]*100.0))
+                print("\tCorr, span: {:3.2f}, {:3.2f}".format(self._errtypes[k]["corr_rec"]*100.0, self._errtypes[k]["span_rec"]*100.0))
+           
+            # Micro F1-score
+            # Results for in-scope categories and all categories
+            if nfreqs != 0:
+                print("Micro F1-score: {:3.2f}  ({:3.2f})".format(micro/nfreqs*100.0, microall/nfreqsall*100.0))
+            else:
+                print("Micro F1-score: N/A")
+
+        def output_supercategory_scores(errorcats: Dict) -> None:
+            # Results for each SÍM category
+            for entry in errorcats:
```

Comment:
Perhaps better: `for entry, catlist in errorcats.items():`

Reply:
Done